[FLINK-14382][yarn] Incorrect handling of FLINK_PLUGINS_DIR on Yarn

wangyang0918 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
This PR is to fix the issue in [FLINK-14382](https://issues.apache.org/jira/browse/FLINK-14382). Flink plugin mechanism could not work for yarn cluster. The plugin directory only needs to be shipped and should not be added to classpath.

After [FLINK-14561](https://issues.apache.org/jira/browse/FLINK-14561), the plugins will be loaded through relative path.


## Brief change log
* Only ship plugin directory on yarn
* Add `AnotherDummyFS` and `WordCount` example to check plugins isolation
* Add e2e test to check plugins isolation for standalone cluster
* Add e2e test to check plugins isolation for yarn cluster


## Verifying this change
E2E tests has been added to check the plugin isolation.

* `./run-single-test.sh test-scripts/test_plugins_isolation.sh`
* `./run-single-test.sh test-scripts/test_plugins_isolation_yarn.sh`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (yes)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -1556,12 +1572,7 @@ public void run() {
-	}
-
-	private void addLibFoldersToShipFiles(Collection<File> effectiveShipFiles) {
+	void addLibFoldersToShipFiles(Collection<File> effectiveShipFiles) {
```
Agree, i will add `@VisibleForTesting`.

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -1580,7 +1591,7 @@ private void addLibFoldersToShipFiles(Collection<File> effectiveShipFiles) {
 	}
 
-	private void addPluginsFoldersToShipFiles(Collection<File> effectiveShipFiles) {
+	void addPluginsFoldersToShipFiles(Collection<File> effectiveShipFiles) {
```
Agree, i will add `@VisibleForTesting`.

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -725,7 +725,10 @@ private ApplicationReport startAppMaster(
 		}
```
It makes sense to me. I will update the commit message to describe the real change.

wangyang0918 commented on flink-end-to-end-tests/flink-plugins-test/wordcount/src/main/java/org/apache/flink/fs/dummy/examples/WordCount.java:
```diff
@@ -0,0 +1,157 @@
+				}
+				Preconditions.checkState(!visible, "%s should not be visible for filesystem %s.",
+					fileSystem1.getClass().getCanonicalName(),
+					fileSystem2.getClass().getCanonicalName());
```
Agree. I will update.

wangyang0918 commented on flink-end-to-end-tests/flink-plugins-test/wordcount/src/main/java/org/apache/flink/fs/dummy/examples/WordCount.java:
```diff
@@ -0,0 +1,157 @@
+		@Override
+		public void flatMap(String value, Collector<Tuple2<String, Integer>> out) throws IOException {
+
+			checkClassIsolation();
```
Agree. I will update.

wangyang0918 commented on flink-end-to-end-tests/flink-plugins-test/wordcount/src/main/java/org/apache/flink/fs/dummy/examples/WordCount.java:
```diff
@@ -0,0 +1,157 @@
+			throw new IllegalArgumentException("--input1 and --input2 should be specified.");
+		}
+		text1 = env.readTextFile(uriOfInput1);
+		text2 = env.readTextFile(uriOfInput2);
```
Agree. I will update.

wangyang0918 commented on flink-end-to-end-tests/flink-plugins-test/another-dummy-fs/pom.xml:
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This commit is usable without next 2 commits. This commit is only to add the another dummy fs and word count example. I will squash them.

wangyang0918 commented:
@AHeise Thanks a lot for your detailed review. I have addressed all your comments. And please take a look.

wangyang0918 commented:
@tillrohrmann Could you please take a look?

tillrohrmann commented:
@wangyang0918 thanks for opening this PR. Unfortunately, I don't have the time to review your PR since I'll be soon on vacation. Maybe @pnowojski can help out with the final verification and merging of the PR.

pnowojski commented on flink-end-to-end-tests/run-pre-commit-tests.sh:
```diff
@@ -54,6 +54,7 @@ run_test "Wordcount end-to-end test" "$END_TO_END_DIR/test-scripts/test_batch_wo
 run_test "Shaded Hadoop S3A end-to-end test" "$END_TO_END_DIR/test-scripts/test_batch_wordcount.sh hadoop"
 run_test "Shaded Presto S3 end-to-end test" "$END_TO_END_DIR/test-scripts/test_batch_wordcount.sh presto"
 run_test "Custom FS plugin end-to-end test" "$END_TO_END_DIR/test-scripts/test_batch_wordcount.sh dummy-fs"
+run_test "Custom FS plugin isolation end-to-end test" "$END_TO_END_DIR/test-scripts/test_plugins_isolation.sh"
```
I don't think we should be adding a separate test for the plugin isolation. It will make the tests unnecessarily execute longer. This new test should supersede the previous `test_bach_wordcount.sh`. 

In other words, `test_batch_wordcount.sh` could be expanded to support multiple (or at least 1 or 2) input files (the expected result would be different for different amount of parameters) and all of the previous usages of the `dummy-fs` should be now using both `dummy-fs` and `another-dummy-fs` file systems.

wangyang0918 commented:
@pnowojski Thanks a lot for your detailed review. I have addressed all your comments.
Update:
* Separate the commits
* Update `ParameterTool` to support multiple value parameter
* Remove old word count and update the existing examples
* Update the e2e test to use two inputs

Please take a look again.

pnowojski commented on flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java:
```diff
@@ -231,13 +253,19 @@ public int getNumberOfParameters() {
+		if (!data.containsKey(key)) {
+			return null;
+		}
+		Preconditions.checkState(data.get(key).size() == 1,
```
nitty nit: maybe add a test case for that?

pnowojski commented on flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java:
```diff
@@ -571,8 +622,15 @@ public ParameterTool mergeWith(ParameterTool other) {
+	 */
 	public Map<String, String> toMap() {
-		return data;
+		return data.entrySet().stream().collect(Collectors.toMap(
```
add `checkState(e.getValue().size() == 1)` and provide a `toMultiMap` variant?

pnowojski commented on flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/wordcount/WordCount.java:
```diff
@@ -99,7 +118,28 @@ public static void main(String[] args) throws Exception {
+				for (String input : inputFiles) {
+					classCollection.add(FileSystem.getUnguardedFileSystem(URI.create(input)));
+				}
+				PluginUtils.checkClassIsolationInDifferentPlugins(classCollection);
```
Those `checkClassIsolationInDifferentPlugins ` (both in batch and streaming `WordCount`) checks  shouldn't happen in the examples, but in the dummy file system implementations. Otherwise they will be confusing people reading those code examples.

wangyang0918 replied:
You concern is reasonable. I will move the class isolation check to dummy file system Implementations.

pnowojski commented on flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java:
```diff
@@ -91,22 +94,23 @@ public static ParameterTool fromArgs(String[] args) {
 
 			i += 1; // try to find the value
 
+			map.putIfAbsent(key, new ArrayList<>());
```
I think by default we should preserve the old behaviour and add an overloaded version of this method `ParameterTool#fromArgs(String[] args, boolean multipleValues)` or something like that?

wangyang0918 replied:
I will try to add a new `MultipleParameterTool extends ParameterTool` instead. It is an enhanced class with multiple parameters support. And it will be marked as `PublicEvolving`. It should be compatible with `ParameterTool`, all the tests of `ParameterToolTest` should pass.

wangyang0918 commented:
@pnowojski Thanks a lot for your detailed review. I have updated the PR. Please take a look again.

Update:
* Re-structure commits and update the commit message.
* Add `MultipleParameterTool ` instead of updating the public API `ParameterTool `
* Move class isolation to dummy fs

pnowojski commented on flink-java/src/main/java/org/apache/flink/api/java/utils/MultipleParameterTool.java:
```diff
@@ -0,0 +1,251 @@
+	}
+
+	// ------------------ ParameterUtil  ------------------------
+	protected final Map<String, Collection<String>> data;
```
Couple of comments here, I think you shouldn't be shading the original `data` variable. If you prefer such form of inheritance (`MultipleParameterTool extends ParameterTool`, as opposed to a composition or some base common class), please rename this `data` -> `multiParameterData` and also make it exclusive. If a parameter is present in one structure, it shouldn't be present in the other. Otherwise it can lead to confusion and bugs.

Couple of methods do not make sense for the `MultiParameterTool`, like `#getConfiguration` or `#getProperties`, as those structures are backed by a regular `Map`. Override them here and throw `UnsupportedOperationException`?

Alternative approach to all of those issues would be to define `AbstractParameterTool`, without a `data` field at all, and with couple of abstract methods, like `#get()`, `#has()` and basically all of the methods that you have implemented in this class as well - those that are different between `MultipleParameterTool` and `ParameterTool`. `ParameterTool` could define `Map<String, String> data` while `MultipleParameterTool` would use `Map<String, Collection<String>> data`, both with appropriate implementation of the abstract methods. 

+ `ParameterTool` would define those methods, that do not make sense for the `MultipleParameterTool` like `#getConfiguration()`.

What do you think?

wangyang0918 replied:
@pnowojski Thanks for your quick response.

I also realize that add `MultipleParameterTool` is an overkill for just one e2e test. However, it is more elegant and will be a valuable addition. 

For now implementation of `MultipleParameterTool`, it is compatible with `ParameterTool`. Even the `#getConfiguration` `#getProperties` could be used. But it is still confusing. So i will follow your suggestion to add `AbstractParameterTool ` with no data. `ParameterTool ` and `MultipleParameterTool ` will have different implementations.

pnowojski replied:
> I also realize that add MultipleParameterTool is an overkill for just one e2e test. However, it is more elegant and will be a valuable addition.

Feel free next time to point out things like this if you realise them regardless if you spot them before/during/after implementing :) It can save us some effort and reviewers are not infallible :(

wangyang0918 replied:
I will point out timely next time :)
Really thanks for your patient review.

wangyang0918 commented:
@pnowojski I have updated the PR just as we have discussed. Only the first commit have changed. Please take a look.

Update:
* Add  `AbstractParameterTool` and move some common method to the abstract class
* Both `ParameterTool` and `MultipleParameterTool` extends the abstract class, they have some different methods.
* Move some common tests to `AbstractParameterToolTest`

pnowojski commented on flink-java/src/main/java/org/apache/flink/api/java/utils/AbstractParameterTool.java:
```diff
@@ -0,0 +1,308 @@
+/**
+ * This class provides common utility methods of {@link ParameterTool} and {@link MultipleParameterTool}.
+ */
+@Public
```
nit: `@Internal`? I guess there is no need to expose this class.

wangyang0918 replied:
The `@Internal` is enough, i will update it.

pnowojski replied:
Ok, let's leave it as it is.

pnowojski commented on flink-java/src/main/java/org/apache/flink/api/java/utils/MultipleParameterTool.java:
```diff
@@ -0,0 +1,269 @@
+	public MultipleParameterTool mergeWith(MultipleParameterTool other) {
+		final Map<String, Collection<String>> resultData = new HashMap<>(data.size() + other.data.size());
+		resultData.putAll(data);
+		resultData.putAll(other.data);
```
what if a key is present both in `this` and `other`? Should we join the collections? 

Frankly If you don't want to spend time fixing it, I would be fine with dropping this method as we do not need it for now. It always can be added in the future. Of course if you would prefer to keep it, I'm also fine with it.

wangyang0918 replied:
Yes, i think we should join the collection. I will fix it quickly.

pnowojski commented:
Travis is green, merging.
